clusterloader2/api/types.go

```diff
@@ -54,20 +54,41 @@ type Config struct {
 	ChaosMonkey ChaosMonkeyConfig `json:"chaosMonkey"`
 }
 
-// Step represents encapsulation of some actions. These actions could be
-// object declarations or measurement usages.
-// Exactly one field (Phases or Measurements) should be non-empty.
+// Step represents a unit of work in ClusterLoader2. It can be either:
+// - a collection of measurements,
+// - a collection of phases,
+// - a group of other steps (module).
+// Exactly one field (Phases or Measurements or Module) should be non-empty.
 type Step struct {
 	// Phases is a collection of declarative definitions of objects.
 	// Phases will be executed in parallel.
 	Phases []Phase `json:"phases"`
 	// Measurements is a collection of parallel measurement calls.
 	Measurements []Measurement `json:"measurements"`
-	// Name is an optional name for given step. If name is set,
-	// timer will be run for the step execution.
+	// Module is a group of steps, usually defined in a separate file.
+	Module Module `json:"module"`
+	// Name is an optional name for given step. If name is set the step execution
+	// time will be measured and the step will be reported at the end of the test.
+	// The name is ignored if the step is of type 'Module'.
 	Name string `json:"name"`
 }
 
+// Module is a group of steps, usually defined in a separate file.
+type Module struct {
+	// Path is the path to the filename with the module template.
+	// If the path is set the the Steps field will get overridden with the content
+	// of the file.
+	Path string `json:"path"`
+	// Params specifies template parameters to be substituted inside the template.
+	Params map[string]interface{} `json:"params"`
+
+	// Steps is the list of steps composing the module. Steps are executed
+	// serially. While it's possible to set the steps directly via this field,
+	// usually a cleaner approach is to have the template definition inside a
+	// separate file and instantiate it via setting the Path/Params fields.
+	Steps []Step `json:"steps"`
```

Comment:
I don't understand this one (and I'm actually against having Steps here).
If you want to define Steps, you can simply define them at the higher level and you don't need module.

Having steps here make it possible to express the same thing in many different ways, which will always be misleading to users.

Let's stick to just a path and params.

Reply:
From a physical point of view Module is a sequence of steps. That's why we have []Steps here. What do you not understand? 

I think the question is whether we should allow initializing modules via setting the Steps directly or just via Path/Params. I don't have a strong opinion here. Current code allows that and it's explained in the comment. 
There might be a value in allowing that in the future. E.g. imagine having conditional modules that would allow to group bunch of steps and disable/enable them depending on some condition. 

So I think it boils down to whether we want to support inline modules (now or in the future). Thoughts?

Reply:
So I understand that 'Steps' field here is an implementation detail and should not be set by the end user?

I can't see any value from 'inline modules', I wouldn't make it possible to set this.

How about removing this field completely?

Reply:
Ok, if two people are baffled by this inline modules then it's clearly not a good idea. I changed the code to not allow inline modules. PTAL